File: RouterLib/src/main/java/com/kronos/router/RouterInject.java
package com.kronos.router;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.text.TextUtils;

import java.util.List;

/**
 * Created by zhangyang on 16/7/18.
 */
public class RouterInject {
    public static void Inject(FragmentActivity activity, Intent intent) {
        Bundle bundle = intent.getExtras();
        String target = bundle != null ? bundle.getString("target") : null;
        IActivityInject inject = getActivityInject(activity);
        if (inject == null) {
            return;
        }
        inject.Inject(bundle);
        FragmentManager fragmentManager = activity.getSupportFragmentManager();
        List<Fragment> fragments = fragmentManager.getFragments();
        for (Fragment fragment : fragments) {
            if (fragment == null) {
                continue;
            }
            IFragmentInject fragmentInject = Inject(fragment, target);
            if (fragmentInject != null) {
                fragmentInject.Inject(bundle);
                break;
            }
        }
    }

    public static void Inject(Fragment fragment, Bundle bundle) {
        String target = bundle != null ? bundle.getString("target") : null;
        IFragmentInject inject = deal(fragment, target);
        if (inject != null) {
            inject.Inject(bundle);
        }
    }

    private static IFragmentInject Inject(Fragment fragment, String target) {
        return deal(fragment, target);
    }

    private static IFragmentInject deal(Object object, String target) {
        if (TextUtils.equals(object.getClass().getName(), target)) {
            return getFragmentInject(object);
        }
        return null;
    }

    private static IActivityInject getActivityInject(Object object) {
        if (object instanceof IActivityInject) {
            return (IActivityInject) object;
        } else {
            return null;
        }
    }

    private static IFragmentInject getFragmentInject(Object object) {
        if (object instanceof IFragmentInject) {
            return (IFragmentInject) object;
        } else {
            return null;
        }
    }

}

File: RouterLib/src/main/java/com/kronos/router/RouterInject.kt
package com.kronos.router

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import androidx.fragment.app.FragmentActivity

/**
 * Created by zhangyang on 16/7/18.
 */
object RouterInject {
    fun inject(activity: FragmentActivity, intent: Intent) {
        val bundle = intent.extras
        val target = bundle?.getString("target")
        val inject = getActivityInject(activity) ?: return
        inject.inject(bundle!!)
        val fragmentManager = activity.supportFragmentManager
        val fragments = fragmentManager.fragments
        for (fragment in fragments) {
            if (fragment == null) {
                continue
            }
            val fragmentInject = inject(fragment, target)
            if (fragmentInject != null) {
                fragmentInject.inject(bundle)
                break
            }
        }
    }

    fun inject(fragment: androidx.fragment.app.Fragment, bundle: Bundle?) {
        val target = bundle?.getString("target")
        val inject = deal(fragment, target)
        inject?.inject(bundle!!)
    }

    private fun inject(fragment: androidx.fragment.app.Fragment, target: String?): IFragmentInject? {
        return deal(fragment, target)
    }

    private fun deal(any: Any, target: String?): IFragmentInject? {
        return if (TextUtils.equals(any.javaClass.name, target)) {
            getFragmentInject(any)
        } else null
    }

    private fun getActivityInject(any: Any): IActivityInject? {
        return if (any is IActivityInject) {
            any
        } else {
            null
        }
    }

    private fun getFragmentInject(any: Any): IFragmentInject? {
        return if (any is IFragmentInject) {
            any
        } else {
            null
        }
    }

}
